File: dotnet/source/amp.eventing/streaming/DefaultEventStream.cs
using amp.bus;
using cmf.bus;
using cmf.eventing.patterns.streaming;
using Common.Logging;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace amp.eventing.streaming
{
    public class DefaultEventStream : IEventStream 
    {
        protected ILog _log;
        private readonly IStandardStreamingEventBus _eventBus;
        private int _batchLimit = 2;
        private Queue<EventStreamQueueItem> _queuedEvents;
        
        private readonly Guid _sequenceId;
        private int _position;
        private string _topic;

        public DefaultEventStream(IStandardStreamingEventBus eventBus, string topic)
        {
            _log = LogManager.GetLogger(this.GetType());
            _eventBus = eventBus;
            _topic = topic;
            _queuedEvents = new Queue<EventStreamQueueItem>();
            _sequenceId = Guid.NewGuid();
            _position = 0;
        }

        public int BatchLimit
        {
            set { _batchLimit = value; }
        }

        public void Publish(object evt)
        {
            _log.Debug("enter publish to stream");
            string sequence = _sequenceId.ToString();

            Envelope env = StreamingEnvelopeHelper.BuildStreamingEnvelope(sequence, _position);
            env.SetMessageTopic(Topic);

            EventContext context = new EventContext(EventContext.Directions.Out, env, evt);
            EventStreamQueueItem eventItem = new EventStreamQueueItem(context);

            _log.Debug("buffering event with sequenceId: " + sequence + ", position: " + _position);
            _queuedEvents.Enqueue(eventItem);

            if (_queuedEvents.Count == (_batchLimit)) {
                _log.Debug("flushing " + _batchLimit + " event(s) to stream.");
                FlushStreamBuffer();
            }
            _position++;
        }

        public string SequenceId
        {
            get { return _sequenceId.ToString(); }
        }

        public string Topic
        {
            get { return _topic; }
        }

        /// <summary>
        /// When processing a stream of an unknown size, it becomes a challenge to know when you have dealt with the 
        /// last object in that stream. This class utilizes the Dispose() method to indicate that stream processing 
        /// has completed.  This is necessary in order to mark the last message with the isLast header flag 
        /// set to true.  The trick here is to ensure that the streamBuffer is not entirely empty when 
        /// dispose gets called.
        /// </summary>
        private void FlushStreamBuffer() 
        {
            while (_queuedEvents.Count > 0) {
                EventStreamQueueItem eventItem = _queuedEvents.Dequeue();
                _eventBus.ProcessEvent(eventItem.EventContext, _eventBus.OutboundProcessors, () =>
                {
                    _eventBus.EnvelopeBus.Send(eventItem.Envelope);
                });
            }
        }

        public void Dispose()
        {
            try
            {
                FlushStreamBuffer();
                _eventBus.Publish(new EndOfStream(Topic, _sequenceId.ToString()));
            }
            catch (Exception e) 
            {
                _log.Error("Unable to send last batch of messages in event stream.", e);
            }
        }




        
    }
}

File: java/amp-eventing/src/main/java/amp/eventing/streaming/DefaultEventStream.java
package amp.eventing.streaming;

import amp.messaging.EnvelopeHelper;
import amp.messaging.MessageContext;
import amp.messaging.IContinuationCallback;
import amp.messaging.MessageException;
import cmf.bus.Envelope;
import cmf.eventing.patterns.streaming.IEventStream;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.*;

public class DefaultEventStream implements IEventStream {
    protected static final Logger log = LoggerFactory.getLogger(DefaultEventStream.class);
    private final IStandardStreamingEventBus eventBus;
    private int batchLimit = 2;
    private Queue<EventStreamQueueItem> queuedEvents;
    private final UUID sequenceId;
    private int position;
    private String topic;

    public DefaultEventStream(IStandardStreamingEventBus eventBus, String topic) {
        this.eventBus = eventBus;
        this.topic = topic;
        this.queuedEvents = new LinkedList<EventStreamQueueItem>();
        this.sequenceId = UUID.randomUUID();
        this.position = 0;
    }

    @Override
    public void setBatchLimit(int numberOfEvents) {
        this.batchLimit = numberOfEvents;
    }

    @Override
    public void publish(Object event) throws Exception {
        log.debug("enter publish to stream");
        String sequence = sequenceId.toString();


        Envelope env = StreamingEnvelopeHelper.buildStreamingEnvelope(sequence, position);
        EnvelopeHelper envHelper = new EnvelopeHelper(env);
        envHelper.setMessageTopic(getTopic());

        MessageContext context = new MessageContext(MessageContext.Directions.Out, event, env);
        EventStreamQueueItem eventItem = new EventStreamQueueItem(context);

        log.debug("buffering event with sequenceId: " + sequence + ", position: " + position);
        this.queuedEvents.add(eventItem);

        if (this.queuedEvents.size() == this.batchLimit) {
            log.debug("flushing " + batchLimit + " event(s) to stream.");
            flushStreamBuffer();
        }

        position++;
    }

    @Override
    public String getTopic() {
        return this.topic;
    }

    @Override
    public String getSequenceId() {
        return this.sequenceId.toString();
    }

    /**
     * When processing a stream of an unknown size, it becomes a challenge to know when you have dealt with the
     * last object in that stream. This class utilizes the dispose() method to indicate that stream processing
     * has completed. This is necessary in order to mark the last message with the isLast flag
     * set to true. The trick here is to ensure that the streamBuffer is not entirely empty when
     * dispose gets called.
     */
    private void flushStreamBuffer() throws Exception {

        while (queuedEvents.size() > 0) {
            final EventStreamQueueItem eventItem = queuedEvents.remove();
            eventBus.processMessage(eventItem.getMessageContext(),
                    new IContinuationCallback() {
                        @Override
                        public void continueProcessing() throws MessageException {
                            try {
								eventBus.getEnvelopeBus().send(eventItem.getEnvelope());
							} catch (Exception ex) {
								log.error("Error sending streaming eventItem.", ex);
								throw new MessageException("Error sending streaming eventItem.", ex);
							}
                        }
                    });
        }
    }

    /**
     * Flushes remaining items in the buffer. This MUST be called to close out the stream or
     * the subscriber will never be notified with the last item in the sequence.
     */
    @Override
    public void dispose() {
        try {
            flushStreamBuffer();
            eventBus.publish(new EndOfStream(this.getTopic(), sequenceId.toString()));

        } catch (Exception e) {
            log.error("Unable to send last batch of messages in buffer to event stream.", e);
        }
    }

}
